trio/_core/tests/test_windows.py

```diff
@@ -58,39 +60,96 @@
-            0,  # no sharing
+            ffi.cast("LPCWSTR", rawname_buf),
+            FileFlags.GENERIC_READ,
+            FileFlags.FILE_SHARE_READ,
             ffi.NULL,  # no security attributes
-            3,  # OPEN_EXISTING
-            0x40000000,  # FILE_FLAG_OVERLAPPED
+            FileFlags.OPEN_EXISTING,
+            FileFlags.FILE_FLAG_OVERLAPPED,
             ffi.NULL,  # no template file
         )
         if handle == INVALID_HANDLE_VALUE:  # pragma: no cover
             raise_winerror()
-        _core.register_with_iocp(handle)
-
-        async def read_region(start, end):
-            await _core.readinto_overlapped(
-                handle,
-                memoryview(buffer)[start:end], start
-            )
 
         try:
-            async with _core.open_nursery() as nursery:
-                for start in range(0, 4096, 512):
-                    nursery.start_soon(read_region, start, start + 512)
-
-            assert buffer == data
-
-            with pytest.raises(TypeError):
-                await _core.readinto_overlapped(handle, b"immutable")
+            with memoryview(buffer) as buffer_view:
+
+                async def read_region(start, end):
+                    await _core.readinto_overlapped(
+                        handle,
+                        buffer_view[start:end],
+                        start,
+                    )
+
+                # Make sure reading before the handle is registered
+                # fails rather than hanging forever
+                with pytest.raises(_core.TrioInternalError) as exc_info:
+                    with move_on_after(0.5):
+                        await read_region(0, 512)
+                assert "Did you forget to call register_with_iocp()?" in str(
+                    exc_info.value
+                )
+
+                _core.register_with_iocp(handle)
+
+                async with _core.open_nursery() as nursery:
+                    for start in range(0, 4096, 512):
+                        nursery.start_soon(read_region, start, start + 512)
+
+                assert buffer == data
+
+                with pytest.raises(BufferError):
+                    await _core.readinto_overlapped(handle, b"immutable")
         finally:
             kernel32.CloseHandle(handle)
 
 
-# XX test setting the iomanager._iocp to something weird to make sure that the
-# IOCP thread can send exceptions back to the main thread
+@contextmanager
+def pipe_with_overlapped_read():
+    from asyncio.windows_utils import pipe
+    import msvcrt
+
+    read_handle, write_handle = pipe(overlapped=(True, False))
+    _core.register_with_iocp(read_handle)
+    try:
+        write_fd = msvcrt.open_osfhandle(write_handle, 0)
+        yield os.fdopen(write_fd, "wb", closefd=False), read_handle
+    finally:
+        kernel32.CloseHandle(ffi.cast("HANDLE", read_handle))
+        kernel32.CloseHandle(ffi.cast("HANDLE", write_handle))
+
+
+async def test_too_late_to_cancel():
```

Comment:
`_core.tests.conftest` doesn't define `--run-slow`, so I'm not sure this will work. Should I copy the `pytest_addoption` method from `tests.conftest`?

Reply:
We should really fix the test layout in general (#274), but yes you can use `@slow` in `_core.tests` (see `test_multierror.py` and `test_ki.py` for examples). The reason it works is that `pytest_addoption` is a global setting, so you only have to call it once per pytest invocation.